<div>
<a href="https://www.audiolabs-erlangen.de/fau/professor/mueller"><img src="data_layout/PCP_Teaser.png" width=100% style="float: right;" alt="PCP Teaser"></a>
</div>

# Unit 3: NumPy Basics

<ul>
    <li><a href='#learn' style="color:#E65100;"><strong>Overview and Learning Objectives</strong></a></li>    
    <li><a href='#background_numpy' style="color:#1B5E20;"><strong>Background:</strong> Why NumPy?</a></li>    
    <li><a href='#module' style="color:black;">Importing and Exploring NumPy</a></li>    
    <li><a href='#numpy' style="color:black;">Creating and Inspecting NumPy Arrays</a></li>
    <li><a href='#reshape' style="color:black;">Reshaping NumPy Arrays</a></li>
    <li><a href='#operation' style="color:black;">Operations on NumPy Arrays</a></li>    
    <li><a href='#type' style="color:black;">NumPy Data Types and Conversion</a></li>    
    <li><a href='#constants' style="color:black;">Mathematical Constants and Special Values</a></li>        
    <li><a href='#exercise_numpy_array' style="color:#006064;"><strong>Exercise 1:</strong>  NumPy Array Manipulations</a></li>
    <li><a href='#background_matrix' style="color:#1B5E20;"><strong>Background:</strong> Matrices and Basic Operations</a></li>        
    <li><a href='#exercise_matrix_operation' style="color:#006064;"><strong>Exercise 2:</strong> Matrix Operations</a></li>        
    <li><a href='#exercise_numpy_math_function' style="color:#006064;"><strong>Exercise 3:</strong> Mathematical Functions</a></li>     
</ul>     




<a id="learn"></a>

<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">

<h2 style="color:#E65100; margin-top:0;">
Overview and Learning Objectives
</h2>

<p style="color:#E65100;">
This unit introduces <a href="https://numpy.org/doc/stable/">NumPy</a>, one of the central packages for scientific computing in Python. Its main data structure is the multidimensional array, which provides an efficient and convenient way to represent numerical data such as vectors, matrices, signals, and images. We begin by creating and inspecting arrays, then explore reshaping, indexing, numerical operations, data types, and special values such as <code>nan</code> and <code>inf</code>. The learning objectives of this unit are as follows:
</p>

<ul style="color:#E65100;">
  <li>Import NumPy and understand the standard abbreviation <code>np</code>.</li>
  <li>Create NumPy arrays and inspect their shape, number of axes, size, and numerical data type.</li>
  <li>Reshape, index, slice, and modify arrays, including the use of Boolean conditions.</li>
  <li>Apply vectorized arithmetic operations and compute sums along selected axes.</li>
  <li>Distinguish element-wise multiplication from matrix multiplication.</li>
  <li>Convert numerical data types and work carefully with special values such as <code>nan</code> and <code>inf</code>.</li>
  <li>Connect the mathematical formulation of vectors and matrices with their implementation in NumPy.</li>
</ul>

<p style="color:#E65100;">
The unit concludes with three programming exercises:
</p>

<ul style="color:#E65100;">
  <li><a href="#exercise_numpy_array">Exercise 1</a>: Create, modify, extend, filter, and sort NumPy arrays.</li>
  <li><a href="#exercise_matrix_operation">Exercise 2</a>: Construct matrices and explore indexing, matrix products, and matrix inversion.</li>
  <li><a href="#exercise_numpy_math_function">Exercise 3</a>: Apply trigonometric, exponential, and rounding functions to numerical arrays.</li>
</ul>

<p style="color:#E65100; margin-bottom:0;">
Before Exercise 2, a <a href="#background_matrix">mathematical recap of matrices and matrix multiplication</a> connects the underlying definitions with their implementation. This interplay between mathematical concepts and executable code is a guiding principle of the PCP notebooks.
</p>

</div>


In [2]:
# --- Core scientific stack ---
import numpy as np

# --- Custom utilities (Unit 3) ---
from libpcp.unit03 import (
    exercise_numpy_array,
    exercise_matrix_operation,
    exercise_numpy_math_function
)

<a id="background_numpy"></a>

<div style="
  background-color: #C8E6C9A0;
  border: 1px solid #81C784A0;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">

<h2 style="
  color:#1B5E20;
  margin-top:0;
  padding-top:0;
">Background: Why NumPy?</h2>

<p style="color:#1B5E20;">
Python is a flexible general-purpose language, but its basic data structures are not designed specifically for large numerical calculations. Scientific applications often process thousands or millions of related values, such as samples of an audio signal, pixels of an image, or measurements from an experiment. NumPy provides a  <strong>data structure </strong> and a  <strong>collection of operations </strong> designed precisely for this kind of work.
</p>

<p style="color:#1B5E20;">
The central NumPy object is the <strong>array</strong>. An array organizes numerical values in a regular structure with one or more <strong>axes</strong>. A one-dimensional array may represent a sequence or vector, a two-dimensional array may represent a table or matrix, and arrays with three or more axes can represent data such as color images, collections of signals, or simulation results. The number of entries along each axis defines the array's <strong>shape</strong>. For example, an array with shape <code>(2, 3)</code> has two rows and three columns. An array is closely related to what many machine-learning frameworks call a <strong>tensor</strong>. In programming, both terms often refer to <strong>multidimensional collections</strong> of numbers. NumPy uses the term array, while frameworks such as PyTorch use tensor for a similar data structure with additional capabilities, including GPU computation and automatic differentiation. In mathematics, however, the word tensor has a more specific meaning and should not always be identified directly with a multidimensional array.
</p>

<p style="color:#1B5E20;">
A Python list is highly flexible and may contain objects of many different types. A <strong>numerical NumPy array</strong> typically stores all its elements using one fixed data type, such as <code>float64</code> or <code>int32</code>, in a compact data buffer. This homogeneous representation generally requires less memory and allows optimized numerical routines to process the data efficiently. Many arrays are stored contiguously, although slicing, transposition, and similar operations may create noncontiguous <strong>views</strong> of existing data. NumPy also supports <strong>vectorized operations</strong>. Instead of writing a Python loop that processes one element at a time, a calculation is expressed for an entire array. For example, <code>x ** 2</code> squares every element of <code>x</code>, while <code>x + 1</code> adds one to every element. The required loops still exist, but they are carried out internally by optimized compiled implementations. Vectorization makes numerical code shorter, easier to read, and often considerably faster.
</p>

<p style="color:#1B5E20; margin-bottom:0;">
<strong>Why NumPy?</strong> It combines the readability of Python with data structures and compiled operations designed for numerical computation. This makes it possible to move naturally from a mathematical expression to an efficient executable experiment. NumPy has therefore become a foundation of the scientific Python ecosystem and is widely used for signals, images, measurements, simulations, and machine-learning data. Further details can be found in the NumPy introduction <a href="https://numpy.org/doc/stable/user/whatisnumpy.html">What is NumPy?</a>.
</p>

</div>


<a id="module"></a>

## Importing and Exploring NumPy

NumPy is a large numerical toolbox, and no one needs to memorize every function it contains. A more useful skill is knowing how to import the package, inspect what it provides, and request help when needed. NumPy is already installed as part of the [PCP environment](PCP_01_getstarted.html). By convention, it is imported using the short name `np`:

In [3]:
import numpy as np

print('NumPy version:', np.__version__)

NumPy version: 2.5.2


The prefix `np` gives access to the objects provided by NumPy. For example, `np.sqrt()` computes square roots, while `np.pi` provides the number $\pi$:

In [4]:
print('Square root of 9:', np.sqrt(9))
print('Pi:', np.pi)

Square root of 9: 3.0
Pi: 3.141592653589793


The function `dir()` returns a sorted Python list of strings containing the names available for an object. When applied to the NumPy module, `dir(np)` lists the names of its functions, classes, constants, submodules, and other attributes. Since NumPy provides hundreds of such names, printing the complete list is not especially helpful. Instead, the list can be inspected or searched:

In [10]:
numpy_names = dir(np)

print('Number of available names:', len(numpy_names))
print('Is sqrt available?', 'sqrt' in numpy_names)
print('Is pi available?', 'pi' in numpy_names)
print('Is fft available?', 'fft' in numpy_names)
print('Is square_root available?', 'square_root' in numpy_names)

Number of available names: 531
Is sqrt available? True
Is pi available? True
Is fft available? True
Is square_root available? False


Python provides several ways to explore a package. The function `dir(np)` returns the available names, while `help()` displays detailed documentation for a selected object. For example, `help(np.sqrt)` describes the NumPy square-root function. Within a Jupyter notebook, the question-mark syntax `np.sqrt?` provides similar information. Since the resulting documentation may be long, we do not display it here. In practice, `dir()` shows what is available, while `help()` and the searchable [NumPy documentation](https://numpy.org/doc/stable/) explain how individual objects work.

<a id="numpy"></a>

## Creating and Inspecting NumPy Arrays

The central NumPy data structure is the **array**. NumPy arrays are objects of the class `numpy.ndarray`, which can also be written as `np.ndarray` after importing NumPy as `np`. They are commonly created using the function `np.array()`.

An array organizes values along one or more axes and typically stores all elements using the same numerical data type. Four properties are particularly important:

* `shape` gives the number of elements along each axis.
* `ndim` gives the number of axes.
* `size` gives the total number of elements.
* `dtype` gives the numerical type used to store the elements.

The following example creates a **one-dimensional array** from a Python list and displays these properties. As introduced in [Unit 2 on Python type conversion](PCP_02_python.html#type), `type()` identifies the type of the complete Python object. For a NumPy array, it returns `numpy.ndarray`. This should not be confused with `dtype`, which describes the numerical type of the elements stored inside the array.

In [15]:
x = np.array([1, 2, 3, 3])

print('Array:', x)
print('Object type:', type(x))
print('Shape:', x.shape)
print('Number of axes:', x.ndim)
print('Number of elements:', x.size)
print('Data type:', x.dtype)

Array: [1 2 3 3]
Object type: <class 'numpy.ndarray'>
Shape: (4,)
Number of axes: 1
Number of elements: 4
Data type: int64


In this example, the shape `(4,)` indicates one axis containing four elements. Note that the **shape** of an array is always represented by a **tuple**. The comma is required for a one-element tuple because Python interprets `(4)` simply as the integer `4`. NumPy also infers the data type from the input values. Here, the result is `int64`, although the default integer type may depend on the computing platform, such as the processor architecture and NumPy build. When a particular type is required, it can be specified explicitly, for example with `np.array([1, 2, 3, 3], dtype=np.int64)`.

In linear algebra, a rectangular arrangement of numbers in rows and columns is called a **matrix**. Such a matrix can be represented in NumPy by a **two-dimensional array**, where the first axis corresponds to the rows and the second axis to the columns. In the following example, the data type is explicitly set to `float64`, so NumPy stores the integer input values as floating-point numbers.

In [24]:
x = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)

print('Two-dimensional array:', x, sep='\n')
print('Shape:', x.shape)
print('Number of axes:', x.ndim)
print('Number of elements:', x.size)
print('Data type:', x.dtype)

Two-dimensional array:
[[1. 2. 3.]
 [4. 5. 6.]]
Shape: (2, 3)
Number of axes: 2
Number of elements: 6
Data type: float64


The shape `(2, 3)` indicates two rows and three columns. More generally, each entry of the shape tuple gives the size of one axis. NumPy provides several convenient functions for creating arrays with particular values or structures:

* `np.zeros(shape)` creates an array filled with zeros.
* `np.ones(shape)` creates an array filled with ones.
* `np.arange(start, stop, step)` creates values separated by a fixed step, with `stop` excluded.
* `np.linspace(start, stop, number)` creates a specified number of evenly spaced values and includes both endpoints by default.
* `np.eye(n)` creates an \(n \times n\) identity matrix, with ones on the main diagonal and zeros elsewhere.

Functions such as `np.zeros()`, `np.ones()`, `np.linspace()`, and `np.eye()` create floating-point arrays by default. This is why values such as `0.` and `1.` contain a decimal point when printed. As discussed above, a different data type can be requested explicitly using `dtype`. The function `np.arange()` infers an integer type when all its arguments are integers.


In [26]:
print('Zeros:', np.zeros(3))
print('Integer zeros:', np.zeros(3, dtype=np.int64))
print('Ones:', np.ones((2, 3)), sep='\n')
print('Values with fixed step:', np.arange(2, 8, 2))
print('Five values from 0 to 1:', np.linspace(0, 1, 5))
print('Identity matrix of size 3:', np.eye(3), sep='\n')

Zeros: [0. 0. 0.]
Integer zeros: [0 0 0]
Ones:
[[1. 1. 1.]
 [1. 1. 1.]]
Values with fixed step: [2 4 6]
Five values from 0 to 1: [0.   0.25 0.5  0.75 1.  ]
Identity matrix of size 3:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


<a id="reshape"></a>
## Reshaping NumPy Arrays

Reshaping changes how the elements of an array are organized along its axes without adding or removing any values. The total number of elements must therefore remain unchanged. For example, the following one-dimensional array contains 24 values, which can be arranged into several different shapes:

In [28]:
x = np.arange(2 * 3 * 4)
print('x =', x)
print('Shape:', x.shape)

y = x.reshape((3, 8))
print('y =', y, sep='\n')
print('Shape:', y.shape)

z = x.reshape((3, 2, 4))
print('z =', z, sep='\n')
print('Shape:', z.shape)

x = [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]
Shape: (24,)
y =
[[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
Shape: (3, 8)
z =
[[[ 0  1  2  3]
  [ 4  5  6  7]]

 [[ 8  9 10 11]
  [12 13 14 15]]

 [[16 17 18 19]
  [20 21 22 23]]]
Shape: (3, 2, 4)


In the previous example, the array `z` has three axes. To access an individual element, one index must be provided for each axis:

In [30]:
print('z[0, 1, 2] =', z[0, 1, 2])

z[0, 1, 2] = 6


NumPy allows one entry of a new shape to be specified as `-1`. This acts as a placeholder whose value is inferred from the number of available elements. For example, an array with six elements can be reshaped to `(3, -1)`, and NumPy determines that the missing value must be `2`.

The shapes `(6,)`, `(6, 1)`, and `(1, 6)` describe different arrays. The first has one axis, while the other two have two axes. In `(6, 1)`, the second axis contains one element; in `(1, 6)`, the first axis contains one element. The position of this singleton axis affects later array and matrix operations.

In [33]:
array_1d = np.arange(6)

array_3x2 = array_1d.reshape((3, -1))
array_6x1 = array_1d.reshape((-1, 1))
array_1x6 = array_1d.reshape((1, -1))

print(f'array_1d:  shape = {array_1d.shape}, axes = {array_1d.ndim}')
print(f'array_3x2: shape = {array_3x2.shape}, axes = {array_3x2.ndim}')
print(f'array_6x1: shape = {array_6x1.shape}, axes = {array_6x1.ndim}')
print(f'array_1x6: shape = {array_1x6.shape}, axes = {array_1x6.ndim}')

array_1d:  shape = (6,), axes = 1
array_3x2: shape = (3, 2), axes = 2
array_6x1: shape = (6, 1), axes = 2
array_1x6: shape = (1, 6), axes = 2


<a id="operation"></a>

## Operations on NumPy Arrays

As discussed in the [background section](#background_numpy), NumPy supports **vectorized operations**. In other words, a single expression can describe a calculation for an entire array without requiring an explicit Python loop over its elements. Many familiar arithmetic and comparison operators can be used in this way. In the following examples, NumPy applies each operation element by element and returns a new array containing the results:

In [34]:
x = np.arange(5)

print('x =', x)
print('x + 1 =', x + 1)
print('x * 2 =', x * 2)
print('x * x =', x * x)
print('x ** 3 =', x ** 3)
print('x / 4 =', x / 4)
print('x // 4 =', x // 4)
print('x > 2 =', x > 2)

x = [0 1 2 3 4]
x + 1 = [1 2 3 4 5]
x * 2 = [0 2 4 6 8]
x * x = [ 0  1  4  9 16]
x ** 3 = [ 0  1  8 27 64]
x / 4 = [0.   0.25 0.5  0.75 1.  ]
x // 4 = [0 0 0 0 1]
x > 2 = [False False False  True  True]


An element-wise operation does not always require two arrays with exactly the same shape. NumPy may automatically adapt a smaller value or array to the required shape, a mechanism called **broadcasting**. The simplest example is `x + 1`: NumPy conceptually applies the scalar value `1` to every element of `x` without creating an equally shaped array of ones. Broadcasting can also combine arrays of different but compatible shapes. In the following example, an array with shape `(3,)` is added to an array with shape `(2, 3)`. NumPy applies the three values to both rows:

In [37]:
x = np.arange(6).reshape((2, 3))
offset = np.array([10, 20, 30])

print('x =', x, sep='\n')
print('offset =', offset)
print('x + offset =', x + offset, sep='\n')

x =
[[0 1 2]
 [3 4 5]]
offset = [10 20 30]
x + offset =
[[10 21 32]
 [13 24 35]]


An important point is that the same operator may behave differently for NumPy arrays and Python lists. For arrays with compatible shapes, `+` performs element-wise addition. For lists, the same operator concatenates the two collections:

In [39]:
array_a = np.array([0, 1, 2, 3])
array_b = np.array([0, 1, 2, 3])
print('Array addition:', array_a + array_b)

list_a = [0, 1, 2, 3]
list_b = [0, 1, 2, 3]
print('List concatenation:', list_a + list_b)

Array addition: [0 2 4 6]
List concatenation: [0, 1, 2, 3, 0, 1, 2, 3]


When two-dimensional arrays represent matrices, the word multiplication can refer to two different operations. The operator `*` multiplies corresponding elements and usually preserves the array shape. The operator `@` computes the mathematical matrix product by combining rows of the first matrix with columns of the second. Element-wise multiplication requires compatible shapes, while matrix multiplication requires the inner dimensions to match:

In [41]:
a = np.arange(4).reshape((2, 2))
b = 2 * np.ones((2, 2), dtype=np.int64)

print('Matrix a:', a, sep='\n')
print('Matrix b:', b, sep='\n')

print('Element-wise sum a + b:', a + b, sep='\n')
print('Element-wise product a * b:', a * b, sep='\n')
print('Matrix product a @ b:', a @ b, sep='\n')

Matrix a:
[[0 1]
 [2 3]]
Matrix b:
[[2 2]
 [2 2]]
Element-wise sum a + b:
[[2 3]
 [4 5]]
Element-wise product a * b:
[[0 2]
 [4 6]]
Matrix product a @ b:
[[ 2  2]
 [10 10]]


Many NumPy operations can combine several elements into a smaller result. Such operations are called **reductions**. As an example, consider the method `sum()`, which adds all elements and returns a single value when no axis is specified. For a two-dimensional array, the summation can also be performed along a selected axis. Summing along `axis=0` combines the rows and produces one value for each column, while summing along `axis=1` combines the columns and produces one value for each row. In each case, the selected axis disappears from the result, as illustrated by the following example:

In [48]:
x = np.arange(6).reshape((2, 3))

sum_axis_0 = x.sum(axis=0)
sum_axis_1 = x.sum(axis=1)

print('Array x:', x, sep='\n')
print('Shape of x:', x.shape)
print('Sum of all elements:', x.sum())
print('Sum along axis 0:', sum_axis_0, 'Shape:', sum_axis_0.shape)
print('Sum along axis 1:', sum_axis_1, 'Shape:', sum_axis_1.shape)

Array x:
[[0 1 2]
 [3 4 5]]
Shape of x: (2, 3)
Sum of all elements: 15
Sum along axis 0: [3 5 7] Shape: (3,)
Sum along axis 1: [ 3 12] Shape: (2,)


NumPy follows the zero-based indexing and slicing conventions introduced for [Python lists](PCP_02_python.html#lists). Index `0` refers to the first element, and the `stop` value of a slice is excluded. For a two-dimensional array, the first index refers to the rows and the second to the columns. A colon `:` selects all elements along the corresponding axis.

A particularly useful form of indexing is based on comparisons. An expression such as `x > 1` creates a Boolean array with the same shape as `x`, containing `True` wherever the condition is satisfied. This Boolean array is called a **mask** and can be used to select the corresponding elements. The following example illustrates indexing, slicing, and masking:

In [50]:
x = np.arange(6).reshape((2, 3))

print('Array x:', x, sep='\n')
print('Element x[1, 1]:', x[1, 1])
print('Slice of first row x[0, 1:3]:', x[0, 1:3])
print('Second row x[1, :]:', x[1, :])
print('Second column x[:, 1]:', x[:, 1])

mask = x > 1
print('Boolean mask x > 1:', mask, sep='\n')
print('Elements selected by the mask:', x[mask])

Array x:
[[0 1 2]
 [3 4 5]]
Element x[1, 1]: 4
Slice of first row x[0, 1:3]: [1 2]
Second row x[1, :]: [3 4 5]
Second column x[:, 1]: [1 4]
Boolean mask x > 1:
[[False False  True]
 [ True  True  True]]
Elements selected by the mask: [2 3 4 5]


<a id="type"></a>
## NumPy Data Types and Conversion

As discussed in [Unit 2 on Python type conversion](PCP_02_python.html#type), Python provides the numerical types `int`, `float`, and `complex`. NumPy provides more specific numerical types whose names often indicate the amount of storage they use. For example, `np.int32` represents a signed integer using 32 bits, while `np.float64` represents a floating-point number using 64 bits.

For an array `x`, the expression `type(x)` identifies the complete object as a NumPy array, while `x.dtype` describes the numerical type of its elements. A particular data type can be selected when an array is created using the argument `dtype`. An existing array can be converted using the method `astype()`, which returns a new array and leaves the original unchanged. The following examples illustrate these conversions:

In [56]:
integer_array = np.array([-1, 2, 3], dtype=np.int64)
floating_array = integer_array.astype(np.float64)
complex_array = integer_array.astype(np.complex128)

print(f'Integer array:        {integer_array}; dtype: {integer_array.dtype}')
print(f'Floating-point array: {floating_array}; dtype: {floating_array.dtype}')
print(f'Complex array:        {complex_array}; dtype: {complex_array.dtype}')

Integer array:        [-1  2  3]; dtype: int64
Floating-point array: [-1.  2.  3.]; dtype: float64
Complex array:        [-1.+0.j  2.+0.j  3.+0.j]; dtype: complex128


The data type determines not only how numbers are stored, but also which results can be represented. For example, the square root of a negative number is not defined within the real numbers. When `np.sqrt()` is applied to a real-valued array containing negative values, NumPy returns `nan` for these entries and issues a `RuntimeWarning`. Here, `nan` means that no valid real-valued result is available.

In [59]:
real_values = np.arange(-2, 2, dtype=np.float64)
real_roots = np.sqrt(real_values)

print(f'Real values: {real_values}; dtype: {real_values.dtype}')
print(f'Square roots: {real_roots}; dtype: {real_roots.dtype}')

Real values: [-2. -1.  0.  1.]; dtype: float64
Square roots: [nan nan  0.  1.]; dtype: float64


C:\Users\meinard\AppData\Local\Temp\ipykernel_24360\1579404477.py:2: RuntimeWarning: invalid value encountered in sqrt
  real_roots = np.sqrt(real_values)


The warning in this example is expected: NumPy is drawing attention to operations that have no valid result in the real number system. This illustrates why choosing a suitable data type matters, since NumPy can only return results within the numerical domain represented by the array. Converting the input to a complex data type extends this domain, allowing the negative values to have valid complex square roots:

In [61]:
complex_values = real_values.astype(np.complex128)
complex_roots = np.sqrt(complex_values)

print(f'Complex values: {complex_values}; dtype: {complex_values.dtype}')
print(f'Square roots: {complex_roots}; dtype: {complex_roots.dtype}')

Complex values: [-2.+0.j -1.+0.j  0.+0.j  1.+0.j]; dtype: complex128
Square roots: [0.+1.41421356j 0.+1.j         0.+0.j         1.+0.j        ]; dtype: complex128


<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">
<p style="color:#E65100;">
<strong>Note: Numerical Warnings.</strong> NumPy normally issues a warning when a calculation produces an exceptional value such as <code>inf</code> or <code>nan</code>. A warning does not necessarily stop the calculation, but it should be examined rather than ignored. If exceptional values are expected and handled deliberately, <a href="https://numpy.org/doc/stable/reference/generated/numpy.errstate.html"><code>np.errstate</code></a> can temporarily suppress selected warnings inside a <code>with</code> block. This changes only the warning behavior, not the values produced by the calculation, and the usual settings are restored automatically when the block ends.
</p>
</div>


<a id="constants"></a>
## Mathematical Constants and Special Values

Numerical calculations often involve well-known mathematical constants or results that cannot be represented as ordinary finite numbers. NumPy provides convenient names for both situations. The first two values below are mathematical constants, while `inf` and `nan` are special floating-point values used to mark exceptional numerical results:

* `np.pi` represents Archimedes' constant $\pi$, the ratio of a circle's circumference to its diameter. It appears throughout geometry, trigonometry, and signal processing.
* `np.e` represents Euler's number $e$, the base of the natural logarithm. It plays a central role in exponential growth, decay, and oscillation.
* `np.inf` and `-np.inf` represent positive and negative infinity. They may arise, for example, when a nonzero floating-point number is divided by zero.
* `np.nan` represents **Not a Number**. It indicates that a numerical result is undefined or unavailable, as in the real-valued square root of a negative number.

In [15]:
print('Pi:', np.pi)
print('Euler’s number:', np.e)
print('Positive infinity:', np.inf)
print('Negative infinity:', -np.inf)
print('Not a Number:', np.nan)

Archimedes constant Pi: 3.141592653589793; type: <class 'float'>
Euler’s constant, base of natural logarithms: 2.718281828459045; type: <class 'float'>
Float representation of (pos.) infinity: inf; type: <class 'float'>
Float representation of (neg.) infinity: -inf; type: <class 'float'>
Float representation of Not a Number (NaN): nan; type: <class 'float'>


Unlike the mathematical constants `np.pi` and `np.e`, the special values `np.inf` and `np.nan` may arise automatically during a calculation. In NumPy floating-point array operations, dividing a positive number by zero produces `inf`, dividing a negative number by zero produces `-inf`, and dividing zero by zero produces `nan`. NumPy continues the calculation while marking the exceptional positions, as illustrated by the following example:

In [64]:
print('10 + inf =', 10 + np.inf)
print('10 + (-inf) =', 10 + (-np.inf))
print('10 + nan =', 10 + np.nan)
print('inf + nan =', np.inf + np.nan)
print('inf - inf =', np.inf - np.inf)

10 + inf = inf
10 + (-inf) = -inf
10 + nan = nan
inf + nan = nan
inf - inf = nan


NumPy provides Boolean tests for identifying special values in an array. The functions `np.isinf()`, `np.isnan()`, and `np.isfinite()` identify infinite, `nan`, and finite values, respectively. Each function returns a Boolean mask with the same shape as the input array. As with the masks introduced earlier, this mask can be used directly to select the corresponding values.

In the following example, several divisions are performed deliberately to generate negative infinity, `nan`, positive infinity, and a finite result. Since the exceptional results are expected, [`np.errstate()`](https://numpy.org/doc/stable/reference/generated/numpy.errstate.html) temporarily suppresses the resulting warnings:

In [72]:
numerator = np.array([-2.0, 0.0, 2.0, 4.0, 6.0, 8.0])
denominator = np.array([0.0, 0.0, 0.0, 0.0, 2.0, 2.0])

with np.errstate(divide='ignore', invalid='ignore'):
    result = numerator / denominator

mask_inf = np.isinf(result)
mask_nan = np.isnan(result)
mask_finite = np.isfinite(result)

print('Result:', result)
print('Infinity mask:', mask_inf)
print('NaN mask:', mask_nan)
print('Finite-value mask:', mask_finite)

print('Infinite values:', result[mask_inf])
print('NaN values:', result[mask_nan])
print('Finite values:', result[mask_finite])

Result: [-inf  nan  inf  inf   3.   4.]
Infinity mask: [ True False  True  True False False]
NaN mask: [False  True False False False False]
Finite-value mask: [False False False False  True  True]
Infinite values: [-inf  inf  inf]
NaN values: [nan]
Finite values: [3. 4.]


<a id='exercise_numpy_array'></a>
<div style="
  background-color: #B2EBF290;
  border: 1px solid #4DD0E190;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">
<strong style="color:#006064; display: block; break-after: avoid; page-break-after: avoid;">Exercise 1: NumPy Array Manipulations</strong>
<ul style="color:#006064;">
    <li> Create a NumPy array <code>a</code> with ascending natural numbers in the interval $[10, 20]=\{10,11,\ldots,20\}$ (using <code>np.arange</code>).</li>
    <li> Set all entries of <code>a</code> to zero where <code>a</code>$\leq13$ and <code>a</code>$>16$.</li>
    <li> Extend the resulting array <code>a</code> with a NumPy array containing the numbers of the interval $[4,6]$ and store the result in a variable <code>b</code> (using <code>np.append</code>).</li>    
    <li> Remove duplicate values of <code>b</code> (using <code>np.unique</code>) and store the result in a variable <code>c</code>. Note that the result is automatically sorted in ascending order.</li>
    <li> Sort <code>c</code> in descending order and store the result in a variable <code>d</code>. Explore and discuss various options to do this including the slicing method <code>c[::-1]</code>, the function <code>reversed()</code>, as well as the NumPy functions <code>np.sort</code> and <code>np.flip</code>.</li>
</ul>    
</div>

In [18]:
# Your Solution

In [19]:
exercise_numpy_array()

[10 11 12 13 14 15 16 17 18 19 20]
[ 0  0  0  0 14 15 16  0  0  0  0]
[ 0  0  0  0 14 15 16  0  0  0  0  4  5  6]
[ 0  4  5  6 14 15 16]
[16 15 14  6  5  4  0] Type: <class 'numpy.ndarray'>
[16 15 14  6  5  4  0] Type: <class 'numpy.ndarray'>
[16 15 14  6  5  4  0] Type: <class 'numpy.ndarray'>
<reversed object at 0x7ae0c823f6d0> Type: <class 'reversed'>
<reversed object at 0x7ae0b821f190> Type: <class 'reversed'>


<a id='background_matrix'></a>
<div style="
  background-color: #C8E6C9A0;
  border: 1px solid #81C784A0;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">
<h2 style="color:#1B5E20;">Background: Matrices and Basic Operations</h2>

<p style="color:#1B5E20;">
Let $N,M\in\mathbb{N}$ be two positive integers. An $(N\times M)$-matrix $\mathbf{A}$ is a rectangular array of entries arranged in <strong>rows</strong> and <strong>columns</strong>. For example, if entries are real numbers, one also writes $\mathbf{A}\in\mathbb{R}^{N\times M}$. Let $a_{nm}\in\mathbb{R}$ denote the entry of $\mathbf{A}$ being in the $n^{\mathrm{th}}$ row and $m^{\mathrm{th}}$ column for $n\in[1:N]=\{1,2,...,N\}$ and $m\in[1:M]$. Then, one also often writes $\mathbf{A}=[a_{nm}]$. A <strong>vector</strong> is a matrix where either $N=1$ (then called <strong>row vector</strong>) or $M=1$ (then called <strong>column vector</strong>).
<br><br>
When computing with vectors and matrices, one has to pay attention that the dimensions $N$ and $M$ of the operands match properly. For example, for a <strong>matrix summation</strong> or <strong>matrix subtraction</strong>, the operands must be of same dimensions or one of them needs to be a scalar (in this case the operations are applied per entry). The multiplication of matrices refers to a <strong>matrix multiplication</strong>. For an $(N\times M)$-matrix $\mathbf{A} = [a_{nm}]$ and a $(M\times P)$-matrix $\mathbf{B} = [b_{mp}]$, the product matrix $\mathbf{C} = \mathbf{A}\mathbf{B}$ with entries $\mathbf{C}=[c_{np}]$ is defined by $c_{np} = \sum_{m=1}^M a_{nm}b_{mp}$ for $n\in[1:N]$ and $p\in[1:P]$. In other words, the entry $c_{np}$ is the <strong>inner product</strong> (sometimes also called <strong>scalar product</strong>) of $n^{\mathrm{th}}$ row of $\mathbf{A}$ and $p^{\mathrm{th}}$ column of $\mathbf{B}$. This calculation rule is illustrated by the following figure:
</p>

<div style="text-align: center;">
    <img src="data/PCP_fig_matmult.svg" width="650px" style="display: block; margin: auto;" alt="PCP_fig_matmult">
</div>
</div>

<a id='exercise_matrix_operation'></a>
<div style="
  background-color: #B2EBF290;
  border: 1px solid #4DD0E190;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">
<strong style="color:#006064; display: block; break-after: avoid; page-break-after: avoid;">Exercise 2: Matrix Operations</strong>
<ul style="color:#006064;">
    <li> Construct a matrix $\mathbf{B} = \begin{bmatrix}2 & 2 & 2 & 2\\2 & 2 & 2 & 2\\0 & 0 & 0 & 0\\\end{bmatrix}$ just using the NumPy functions <code>np.zeros</code>, <code>np.ones</code>, and <code>np.vstack</code>. Check the matrix dimensions using <code>np.shape</code>.</li>
    <li> Find the row and column index of the maximum entry of the matrix $\mathbf{D} = \begin{bmatrix}2 & 0 & 2\\-1 & 5 & 10\\-3 & 0 & 9\\\end{bmatrix}$ using the functions <code>np.max</code>, <code>np.argmax</code> and <code>np.unravel_index</code>. Why is it not sufficient to use <code>np.argmax</code>?</li>
    <li> Given a row vector $\mathbf{v} = [3\;2\;1]$ and a column vector $\mathbf{w} = [6\;5\;4]^T$, compute $\mathbf{v}\mathbf{w}$ and $\mathbf{w}\mathbf{v}$ using <code>np.dot</code> and <code>np.outer</code>. What is the difference between <code>np.multiply</code>, <code>np.dot</code>, and <code>np.outer</code>?</li>
    <li> Given $\mathbf{A} = \begin{bmatrix}1 & 2\\3 & 5\end{bmatrix}$, $\mathbf{v} = \begin{bmatrix}1\\4\end{bmatrix}$, compute  $\mathbf{A}^{-1}$ and $\mathbf{A}^{-1}\mathbf{v}$ (using <code>np.linalg.inv</code>).</li>
    </ul>
</div>

In [20]:
# Your Solution

In [21]:
exercise_matrix_operation()

Matrix B: 
[[2. 2. 2. 2.]
 [2. 2. 2. 2.]
 [0. 0. 0. 0.]]
Shape of matrix B: (3, 4)
Maximum of matrix D: 10
Row and column index of maximum entry: (np.int64(1), np.int64(2))
vw =  32
wv = 
[[18 12  6]
 [15 10  5]
 [12  8  4]]
Inverse(A) = 
[[-5.  2.]
 [ 3. -1.]]
Inverse(A)v = 
[ 3. -1.]


<a id='exercise_numpy_math_function'></a>
<div style="
  background-color: #B2EBF290;
  border: 1px solid #4DD0E190;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 20px;
  padding-right: 20px;
  margin-top: 10px;
  margin-bottom: 10px;
">
<strong style="color:#006064; display: block; break-after: avoid; page-break-after: avoid;">Exercise 3: Mathematical NumPy Functions</strong>

<p style="color:#006064;">
The NumPy package offers many different <a href='https://numpy.org/doc/stable/reference/routines.math.html'>mathematical functions</a>. Explore these functions by trying them out on small examples. In particular, you should gain a good understanding of the following concepts.
</p>
<ul style="color:#006064;">
    <li>Generate a NumPy array <code>v_deg</code> containing the numbers $[0, 30, 45, 60, 90, 180]$. Apply the function <code>np.deg2rad</code> to convert this array (given in degree) to an array <code>v_rad</code> (given in radiants). Then apply the functions <code>np.cos</code> and <code>np.sin</code> and discuss the results.</li>
    <li>Using the the same array as before, apply the exponential function  <code>np.exp(1j * v_rad)</code>. What is meant by <code>1j</code>? What is the relation to <code>np.cos</code> and <code>np.sin</code>? Use the functions <code>np.real</code>, <code>np.imag</code>, and <code>np.isclose</code> to make this relation explicit.
    <li>Try out different functions for rounding using the numbers $[-3.1416, -1.5708, 0, 1.5708, 3.1416]$. What is the difference between the functions <code>np.round</code>, <code>np.floor</code>, <code>np.ceil</code>, and <code>np.trunc</code>? </li>  
</ul>    
</div>

In [22]:
# Your Solution

In [23]:
exercise_numpy_math_function()


 ==== Cos and sin function ====
deg:  [  0  30  45  60  90 180]
rad:  [ 0.0000  0.5236  0.7854  1.0472  1.5708  3.1416]
cos:  [ 1.0000  0.8660  0.7071  0.5000  0.0000 -1.0000]
sin:  [ 0.0000  0.5000  0.7071  0.8660  1.0000  0.0000]

 ==== Exponential function ====
exp:  [ 1.00000000e+00+0.00000000e+00j  8.66025404e-01+5.00000000e-01j
  7.07106781e-01+7.07106781e-01j  5.00000000e-01+8.66025404e-01j
  6.12323400e-17+1.00000000e+00j -1.00000000e+00+1.22464680e-16j]
exp_real:  [ 1.0000  0.8660  0.7071  0.5000  0.0000 -1.0000]
exp_imag:  [ 0.0000  0.5000  0.7071  0.8660  1.0000  0.0000]
cos == exp_real: [ True  True  True  True  True  True]
sin == exp_imag: [ True  True  True  True  True  True]

 ==== Rounding options ====
Original numbers:           [-3.1416 -1.5708  0.0000  1.5708  3.1416]
Round to integer:           [-3.0000 -2.0000  0.0000  2.0000  3.0000]
Round to three decimals:    [-3.1420 -1.5710  0.0000  1.5710  3.1420]
Return floor of numbers:    [-4.0000 -2.0000  0.0000  1.0000 

<div>
<a href="https://opensource.org/licenses/MIT"><img src="data_layout/PCP_License.png" width=100% style="float: right;" alt="PCP License"></a>
</div>